Disallow engine change on sample db

escherize opened the pull request:
Fixes [#16382](https://github.com/metabase/metabase/issues/16382). 

In this PR, we do not allow changes to the `engine` of sample databases (any Database where `is_sample` is `true`). This work does not gray out fields, which turned out to be non-trivial, and involved implementing a read only mode for our Select component. This PR should be followed on with changes to make the engine select be disabled.

escherize committed:
for sample db do not allow certain updates

- namely to details or engine in the model layer

escherize committed:
Merge branch 'master' into disallow-changing-details+engine-on-sample-db

escherize committed:
allow sample db updates on start

- update copy

escherize committed:
improve docstring

escherize committed:
cleanup

escherize committed:
more cleanup

escherize committed:
fixing tests

escherize committed:
Merge branch 'master' into disallow-changing-details+engine-on-sample-db

escherize committed:
offer an escape hatch to force edit sample db

- use this hatch from e2e tests

escherize committed:
force_sample goes in request.body

- not in request.body.details

escherize committed:
Revert "offer an escape hatch to force edit sample db"

This reverts commit e2ec947efa2b9d760fec6c6b09098023da481d95.

escherize committed:
remove update details constraint

escherize committed:
remove *allow-sample-update?* dynamic var

- no longer needed

escherize committed:
more cleanup

escherize committed:
only check fornew schedules onc time

noahmoss commented on src/metabase/models/database.clj:
```diff
@@ -135,36 +135,48 @@
+                                                         :name
+                                                         :is_sample] :id (u/the-id database))
+        new-engine                       (some-> new-engine keyword)]
+    (if (and is-sample?
```
nit: up to you, but you could use `when` here to avoid indenting the rest of the function since the exception will short-circuit it

camsaul replied:
There are pros and cons to both approaches. I tend to lean on the side of trying to make choices that minimize the overall indentation level of the code, e.g. I personally find flatter code like

```c++
if (...) {
  return;
}
DoSomethingElse();
```

easier to read than 

```c++
if (...) {
  return;
} else {
  DoSomethingElse();
}
```

[Someone else put the reasoning for this better than me](http://blog.wilshipley.com/2005/07/code-insults-mark-i.html?q=code+insults):

> Don't indent and indent and indent for the main flow of the method. This is huge. Most people learn the exact opposite way from what's really proper — they test for a correct condition, and if it's true, they continue with the real code inside the "if".
>
> What you should really do is write "if" statements that check for improper conditions, and if you find them, bail. This cleans your code immensely, in two important ways: (a) the main, normal execution path is all at the top level, so if the programmer is just trying to get a feel for the routine, all she needs to read is the top level statements, instead of trying to trace through indention levels figuring out what the "normal" case is, and (b) it puts the "bail" code right next to the correctness check, which is good because the "bail" code is usually very short and belongs with the correctness check.

Just my 2 cents tho. I just wanted to point out good reasons you might consider keeping things flat as well

noahmoss commented on test/metabase/models/database_test.clj:
```diff
@@ -216,6 +216,24 @@
                       (format "Secret ID %d was not removed from the app DB" secret-id)))))))))))
 
+(deftest user-may-not-update-sample-database-test
+  (binding [api/*current-user-id* (mt/user->id :crowberto)]
```
is this binding important for the test?

if you do need it, you can use the `with-test-user` macro instead

escherize replied:
Good catch! I don't think the user is related to the test at all.

escherize committed:
remove unnecessary *current-user-id* binding

- as per code review

camsaul commented on test/metabase/models/database_test.clj:
```diff
@@ -216,6 +216,23 @@
+                                                                    :name      "Sample Database"
+                                                                    :details   {:db "./resources/sample-database.db;USER=GUEST;PASSWORD=guest"}}]
+    (testing " updating the engine of a sample database is not allowed"
+      (try (db/update! Database id :engine :sqlite)
+           (catch Exception e
+             (is (= "The engine on a sample database cannot be changed." (.getMessage e)))
+             (is (= {:status-code     400
+                     :existing-engine :h2
+                     :new-engine      :sqlite}
+                    (ex-data e))))))
```
You should do `(is (thrown-with-msg? ...))` here because if this **does not** throw an Exception then those `is` forms inside the `catch` are never going to run. `thrown?` and `throw-with-msg?` actually assert that an Exception is thrown and fail if one is not

camsaul commented:
Did we still want to disallow changing `details` in the API (but still allow it for the special case where we update it on launch)?

escherize commented:
Reviewing the issue, changing `details` did not cause our users any problems. Also that is the piece that was complicating our tests so I decided to leave it off. It simplified this submission a lot.

escherize committed:
Use `thrown-with-message?` to check for exceptions

- Previously this would pass if db/update did _not_ throw an excetption,
  which is not what we want here.

camsaul reviewed (approved):
Approving with the caveat that you should change that test to use `thrown-with-msg?` so those specific assertions fail if an Exception isn't thrown

escherize commented:
So, we just allow it to be updated all the time, and if someone wants to change their sample db details, e.g. syncing schedule, well that isn't going to cause any connection problems for them.

escherize commented:
> the caveat that you should change that test to use thrown-with-msg? so those specific assertions fail if an Exception isn't thrown

Will do!
